File: dotnet/src/SimpleOrm/SchemaGuard.cs
using System.Data.Common;
using System.Reflection;
using System.Text.RegularExpressions;

namespace SimpleOrm;

/// <summary>
/// The rules (§7.18–21): validates every registered query/command and every mapped
/// entity against the real database **without executing anything that writes** —
/// statements are prepared and described inside a transaction that is always rolled
/// back. Every violation is collected; one <see cref="SchemaValidationException"/>
/// carries the complete report. Runs at startup and in a test that calls the same
/// code; there is no warn-only mode.
/// </summary>
public static class SchemaGuard
{
    private static readonly Regex SelectStar = new(
        @"(?i)(?<=select|,)\s*([A-Za-z_]\w*\s*\.\s*)?\*", RegexOptions.Compiled);

    private static readonly Regex NotNullComment = new(
        @"(?i)--\s*notnull:\s*([^\r\n]+)", RegexOptions.Compiled);

    /// <summary>Validates the assembly's registries, entities, and migration state against the session's database.</summary>
    public static Task ValidateAsync(Db db, Assembly assembly, CancellationToken ct)
        => ValidateAsync(db, assembly, migrationsNamespace: null, ct);

    public static async Task ValidateAsync(Db db, Assembly assembly, string? migrationsNamespace, CancellationToken ct)
    {
        var errors = new List<ValidationError>();
        await CheckMigrationsAsync(db, assembly, migrationsNamespace, errors, ct).ConfigureAwait(false);
        await ValidateTypesCoreAsync(db, assembly.GetExportedTypes(), errors, ct).ConfigureAwait(false);
        Throw(errors);
    }

    /// <summary>Validates an explicit set of registry classes and entity types (test harnesses, partial checks).</summary>
    public static async Task ValidateTypesAsync(Db db, IEnumerable<Type> types, CancellationToken ct)
    {
        var errors = new List<ValidationError>();
        await ValidateTypesCoreAsync(db, types.ToArray(), errors, ct).ConfigureAwait(false);
        Throw(errors);
    }

    private static void Throw(List<ValidationError> errors)
    {
        if (errors.Count > 0)
        {
            throw new SchemaValidationException(errors);
        }
    }

    private static async Task ValidateTypesCoreAsync(
        Db db, IReadOnlyList<Type> types, List<ValidationError> errors, CancellationToken ct)
    {
        // The rollback shield: nothing validated here can leave a trace.
        using var shield = db.Connection.BeginTransaction();
        var cache = new Dictionary<string, Dictionary<string, ColumnInfo>>(StringComparer.OrdinalIgnoreCase);
        try
        {
            foreach (var (source, sql, argsType, resultType) in DiscoverRegistry(types))
            {
                await ValidateStatementAsync(db, shield, cache, source, sql, argsType, resultType, errors, ct).ConfigureAwait(false);
            }

            foreach (var entity in types.Where(t =>
                t is { IsClass: true, IsAbstract: false } && EntityMapLoader.HasMappingAttributes(t)))
            {
                await ValidateEntityAsync(db, shield, cache, entity, errors, ct).ConfigureAwait(false);
            }
        }
        finally
        {
            shield.Rollback();
        }
    }

    // --- registry ------------------------------------------------------------------

    private static IEnumerable<(string Source, string Sql, Type? ArgsType, Type? ResultType)> DiscoverRegistry(
        IReadOnlyList<Type> types)
    {
        foreach (var type in types)
        {
            foreach (var field in type.GetFields(BindingFlags.Public | BindingFlags.Static))
            {
                if (!field.FieldType.IsGenericType)
                {
                    continue;
                }

                var definition = field.FieldType.GetGenericTypeDefinition();
                var source = $"{type.Name}.{field.Name}";
                var value = field.GetValue(null);
                if (value is null)
                {
                    continue;
                }

                var arguments = field.FieldType.GetGenericArguments();
                if (definition == typeof(Query<,>))
                {
                    yield return (source, SqlOf(value), arguments[0], arguments[1]);
                }
                else if (definition == typeof(Command<>))
                {
                    yield return (source, SqlOf(value), arguments[0], null);
                }
            }
        }

        static string SqlOf(object entry)
            => ((SqlSource)entry.GetType().GetProperty("Source")!.GetValue(entry)!).Sql;
    }

    private static async Task ValidateStatementAsync(
        Db db, DbTransaction shield, Dictionary<string, Dictionary<string, ColumnInfo>> cache, string source, string sql,
        Type? argsType, Type? resultType, List<ValidationError> errors, CancellationToken ct)
    {
        var placeholders = SqlPlaceholders.Find(sql);

        // PRM-001/PRM-002 statically, both directions (§7.13).
        if (argsType is not null)
        {
            var properties = argsType.GetProperties(BindingFlags.Public | BindingFlags.Instance)
                .Where(p => p.GetMethod is { IsPublic: true })
                .Select(p => p.Name)
                .ToArray();
            foreach (var placeholder in placeholders.Where(p =>
                !properties.Any(n => string.Equals(n, p, StringComparison.OrdinalIgnoreCase))))
            {
                errors.Add(new ValidationError(
                    "PRM-001", source, $"SQL parameter @{placeholder} has no property on {argsType.Name}"));
            }

            foreach (var property in properties.Where(n =>
                !placeholders.Any(p => string.Equals(n, p, StringComparison.OrdinalIgnoreCase))))
            {
                errors.Add(new ValidationError(
                    "PRM-002", source, $"property {argsType.Name}.{property} is never used by the SQL"));
            }
        }

        // Lints on the raw text.
        var stripped = StripLiteralsAndComments(sql);
        if (SelectStar.IsMatch(stripped))
        {
            errors.Add(new ValidationError("VAL-021", source, "SELECT * is not allowed; list columns explicitly"));
        }

        if (Regex.IsMatch(stripped, @"(?i)current_timestamp|datetime\s*\(\s*'now'"))
        {
            errors.Add(new ValidationError(
                "VAL-020", source, "current_timestamp/datetime('now') store datetimes without a UTC marker; bind an ISO-8601 Z value instead"));
        }

        // Prepare / describe without executing anything that persists.
        using var command = db.Connection.CreateCommand();
        command.Transaction = shield;
        command.CommandText = sql;
        foreach (var placeholder in placeholders)
        {
            var parameter = command.CreateParameter();
            parameter.ParameterName = "@" + placeholder;
            parameter.Value = DBNull.Value;
            command.Parameters.Add(parameter);
        }

        if (resultType is null)
        {
            try
            {
                command.Prepare();
            }
            catch (DbException exception)
            {
                errors.Add(new ValidationError("VAL-001", source, exception.Message));
            }

            return;
        }

        DbDataReader reader;
        try
        {
            reader = await command.ExecuteReaderAsync(ct).ConfigureAwait(false);
        }
        catch (DbException exception)
        {
            errors.Add(new ValidationError("VAL-001", source, exception.Message));
            return;
        }

        try
        {
            // Result shape (MAP-001/002/003) through the one mapping pipeline.
            try
            {
                typeof(ResultMapper).GetMethod(nameof(ResultMapper.CreatePlan))!
                    .MakeGenericMethod(resultType)
                    .Invoke(db.Mapper, [reader, source]);
            }
            catch (TargetInvocationException exception) when (exception.InnerException is SimpleOrmException mapping)
            {
                errors.Add(new ValidationError(mapping.Code, source, mapping.Message));
            }

            await ValidateResultColumnsAsync(db, shield, cache, source, sql, reader, resultType, errors, ct).ConfigureAwait(false);
        }
        finally
        {
            reader.Dispose();
        }
    }

    private static async Task ValidateResultColumnsAsync(
        Db db, DbTransaction shield, Dictionary<string, Dictionary<string, ColumnInfo>> cache, string source, string sql,
        DbDataReader reader, Type resultType, List<ValidationError> errors, CancellationToken ct)
    {
        var notNullOverrides = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        foreach (Match match in NotNullComment.Matches(sql))
        {
            foreach (var name in match.Groups[1].Value.Split(','))
            {
                notNullOverrides.Add(name.Trim());
            }
        }

        var schema = TryGetSchemaTable(reader);
        var entityMap = EntityMapLoader.HasMappingAttributes(resultType) ? db.Maps.Load(resultType) : null;

        for (var i = 0; i < reader.FieldCount; i++)
        {
            var column = reader.GetName(i);
            var member = ResolveMember(resultType, entityMap, column);
            if (member is null)
            {
                continue;   // MAP-001 already reported by the pipeline
            }

            var (memberType, memberNullable) = member.Value;
            var origin = OriginOf(schema, i);
            if (origin is { } o)
            {
                var info = await GetColumnInfoAsync(db, shield, cache, o.Table, o.Column, ct).ConfigureAwait(false);
                if (info is { } columnInfo)
                {
                    var enumAsInt = entityMap?.Properties
                        .FirstOrDefault(p => string.Equals(p.ColumnName, column, StringComparison.OrdinalIgnoreCase))
                        ?.EnumAsInt ?? false;
                    if (!db.Options.Dialect.IsDeclaredTypeCompatible(columnInfo.DeclaredType, memberType, enumAsInt)
                        && !db.Converter.HasHandler(Nullable.GetUnderlyingType(memberType) ?? memberType))
                    {
                        errors.Add(new ValidationError(
                            "VAL-011", source,
                            $"column '{column}' is declared {columnInfo.DeclaredType} in {o.Table}, incompatible with {memberType.Name} (no handler)"));
                    }

                    if (!columnInfo.NotNull && !memberNullable)
                    {
                        errors.Add(new ValidationError(
                            "VAL-010", source, $"nullable column '{column}' maps to non-nullable {memberType.Name}"));
                    }

                    continue;
                }
            }

            // Expression column: nullability unknowable — require nullable or the comment.
            if (!memberNullable && !notNullOverrides.Contains(column))
            {
                errors.Add(new ValidationError(
                    "VAL-010", source,
                    $"expression column '{column}' has unknown nullability; make the member nullable or add '-- notnull: {column}'"));
            }
        }
    }

    // --- entities ------------------------------------------------------------------

    private static async Task ValidateEntityAsync(
        Db db, DbTransaction shield, Dictionary<string, Dictionary<string, ColumnInfo>> cache, Type entityType,
        List<ValidationError> errors, CancellationToken ct)
    {
        EntityMap map;
        try
        {
            map = db.Maps.Load(entityType);
        }
        catch (MappingException exception)
        {
            errors.AddRange(exception.Errors.Select(e => new ValidationError(e.Code, entityType.Name, e.Message)));
            return;
        }

        switch (map.Kind)
        {
            case RelationKind.Procedure when !db.Options.Dialect.SupportsProcedures:
            case RelationKind.MaterializedView when !db.Options.Dialect.SupportsMaterializedViews:
                return;   // dormant on this dialect (capability-gated)
            case RelationKind.Statement:
                await ValidateStatementAsync(
                    db, shield, cache, entityType.Name + " [Statement]", map.DefiningSql!,
                    argsType: null, resultType: entityType, errors, ct).ConfigureAwait(false);
                return;
        }

        var columns = await GetRelationColumnsAsync(db, shield, map.RelationName!, ct).ConfigureAwait(false);
        if (columns.Count == 0)
        {
            errors.Add(new ValidationError(
                "VAL-012", entityType.Name, $"relation '{map.RelationName}' does not exist in the database"));
            return;
        }

        foreach (var property in map.Properties)
        {
            if (!columns.TryGetValue(property.ColumnName, out var info))
            {
                errors.Add(new ValidationError(
                    "VAL-013", entityType.Name, $"mapped column '{property.ColumnName}' does not exist in '{map.RelationName}'"));
                continue;
            }

            if (map.Kind != RelationKind.Table)
            {
                continue;   // views report neither declared types nor nullability reliably (§7.19)
            }

            if (!db.Options.Dialect.IsDeclaredTypeCompatible(info.DeclaredType, property.ClrType, property.EnumAsInt)
                && !db.Converter.HasHandler(Nullable.GetUnderlyingType(property.ClrType) ?? property.ClrType))
            {
                errors.Add(new ValidationError(
                    "VAL-011", entityType.Name,
                    $"column '{property.ColumnName}' is declared {info.DeclaredType}, incompatible with {property.ClrType.Name} (no handler)"));
            }

            if (!info.NotNull && !property.IsNullable)
            {
                errors.Add(new ValidationError(
                    "VAL-010", entityType.Name,
                    $"nullable column '{property.ColumnName}' maps to non-nullable {property.PropertyName}"));
            }
        }
    }

    // --- migrations (MIG-030 and history health) ------------------------------------

    private static async Task CheckMigrationsAsync(
        Db db, Assembly assembly, string? migrationsNamespace, List<ValidationError> errors, CancellationToken ct)
    {
        MigrationRunner runner;
        try
        {
            runner = new MigrationRunner(db, assembly, migrationsNamespace);
        }
        catch (SimpleOrmException exception)
        {
            errors.Add(new ValidationError(exception.Code, "migrations", exception.Message));
            return;
        }

        foreach (var entry in await runner.StatusAsync(ct).ConfigureAwait(false))
        {
            var (code, message) = entry.State switch
            {
                MigrationState.Pending => ("MIG-030", "pending — apply migrations before starting the application"),
                MigrationState.Drifted => ("MIG-010", "applied with a different checksum than the code renders"),
                MigrationState.Unknown => ("MIG-011", "applied in the database but unknown to the code"),
                _ => (null as string, string.Empty),
            };
            if (code is not null)
            {
                errors.Add(new ValidationError(code, $"V{entry.Version:0000} {entry.ObjectName}", message));
            }
        }
    }

    // --- introspection helpers ------------------------------------------------------

    private sealed record ColumnInfo(string DeclaredType, bool NotNull);

    private static async Task<Dictionary<string, ColumnInfo>> GetRelationColumnsAsync(
        Db db, DbTransaction shield, string relation, CancellationToken ct)
    {
        var columns = new Dictionary<string, ColumnInfo>(StringComparer.OrdinalIgnoreCase);
        using var command = db.Connection.CreateCommand();
        command.Transaction = shield;
        command.CommandText = db.Options.Dialect.ColumnsInfoSql;
        var parameter = command.CreateParameter();
        parameter.ParameterName = "@relation";
        parameter.Value = relation;
        command.Parameters.Add(parameter);

        var reader = await command.ExecuteReaderAsync(ct).ConfigureAwait(false);
        try
        {
            while (await reader.ReadAsync(ct).ConfigureAwait(false))
            {
                // Primary-key columns are implicitly NOT NULL (SQLite reports notnull=0 for INTEGER PRIMARY KEY).
                var notNull = (long)reader.GetValue(2) != 0 || (long)reader.GetValue(3) != 0;
                columns[reader.GetString(0)] = new ColumnInfo(reader.GetString(1), notNull);
            }
        }
        finally
        {
            reader.Dispose();
        }

        return columns;
    }

    private static async Task<ColumnInfo?> GetColumnInfoAsync(
        Db db, DbTransaction shield, Dictionary<string, Dictionary<string, ColumnInfo>> cache,
        string table, string column, CancellationToken ct)
    {
        if (!cache.TryGetValue(table, out var columns))
        {
            columns = await GetRelationColumnsAsync(db, shield, table, ct).ConfigureAwait(false);
            cache[table] = columns;
        }

        return columns.TryGetValue(column, out var info) ? info : null;
    }

    private static (string Table, string Column)? OriginOf(System.Data.DataTable? schema, int ordinal)
    {
        if (schema is null || ordinal >= schema.Rows.Count)
        {
            return null;
        }

        var row = schema.Rows[ordinal];
        var table = Value(row, "BaseTableName");
        var column = Value(row, "BaseColumnName");
        return table is null || column is null ? null : (table, column);

        static string? Value(System.Data.DataRow row, string name)
            => row.Table.Columns.Contains(name) && row[name] is string s && s.Length > 0 ? s : null;
    }

    private static System.Data.DataTable? TryGetSchemaTable(DbDataReader reader)
    {
        try
        {
            return reader.GetSchemaTable();
        }
        catch (Exception)
        {
            return null;
        }
    }

    private static (Type Type, bool Nullable)? ResolveMember(Type resultType, EntityMap? map, string column)
    {
        if (map is not null)
        {
            var mapped = map.Properties.FirstOrDefault(
                p => string.Equals(p.ColumnName, column, StringComparison.OrdinalIgnoreCase));
            return mapped is null ? null : (mapped.ClrType, mapped.IsNullable);
        }

        if (IsScalarish(resultType))
        {
            return (resultType, !resultType.IsValueType || Nullable.GetUnderlyingType(resultType) is not null);
        }

        var property = resultType.GetProperties(BindingFlags.Public | BindingFlags.Instance)
            .FirstOrDefault(p => NamesMatch(column, p.Name));
        if (property is not null)
        {
            return (property.PropertyType, NullabilityReader.IsNullable(property));
        }

        return null;
    }

    private static bool IsScalarish(Type type)
    {
        var underlying = Nullable.GetUnderlyingType(type) ?? type;
        return underlying.IsPrimitive || underlying.IsEnum || underlying == typeof(string)
            || underlying == typeof(decimal) || underlying == typeof(Guid)
            || underlying == typeof(DateTime) || underlying == typeof(DateTimeOffset) || underlying == typeof(byte[]);
    }

    private static bool NamesMatch(string column, string member)
        => string.Equals(
            column.Replace("_", string.Empty),
            member.Replace("_", string.Empty),
            StringComparison.OrdinalIgnoreCase);

    private static string StripLiteralsAndComments(string sql)
        => Regex.Replace(
            Regex.Replace(sql, @"'([^']|'')*'", "''"),
            @"--[^\r\n]*", string.Empty);
}

File: go/orm/schema_guard.go
package orm

// SchemaGuard is the startup rules (§7.18-§7.21, mirrors
// dotnet/src/SimpleOrm/SchemaGuard.cs and php/src/Validation/SchemaGuard.php):
// [Validate] checks everything a [Registry] declares against the real
// database the session is connected to — every registered query/command,
// every mapped entity, and, when declared, the migration history — without
// executing anything that writes. Every violation is collected into one
// *core.ValidationErrors ([ValidationErrors]); there is no first-error-only
// and no warn-only mode. SchemaGuard lives in this package (not a sibling
// package) so it can reach the session's connection, loader, converter, and
// mapper directly (CODING-STANDARD §1) — the one mapping pipeline
// (db.mapper) is reused rather than built twice.
//
// Go has no field names for a package-level query/command value the way C#
// has Type.Field, so a registry entry's source is its SQLSource.Description()
// — the file path, or a prefix of the inline SQL (CODING-STANDARD §10).

import (
	"context"
	"errors"
	"fmt"
	"reflect"
	"regexp"
	"strings"

	"github.com/kosmassp/SimpleOrm/go/orm/internal/core"
	"github.com/kosmassp/SimpleOrm/go/orm/internal/mapping"
	"github.com/kosmassp/SimpleOrm/go/orm/internal/metadata"
	"github.com/kosmassp/SimpleOrm/go/orm/migrations"
)

var (
	// selectStarPattern ports the C# reference's lookbehind regex
	// `(?<=select|,)\s*([A-Za-z_]\w*\s*\.\s*)?\*` to Go's RE2, which has no
	// lookbehind: the alternative keeps "select"/"," inside the match instead
	// (CODING-STANDARD §10 — VAL-021 only cares whether it matches).
	selectStarPattern     = regexp.MustCompile(`(?i)(?:select|,)\s*(?:[A-Za-z_]\w*\s*\.\s*)?\*`)
	nonUTCTimestampSQL    = regexp.MustCompile(`(?i)current_timestamp|datetime\s*\(\s*'now'`)
	notNullCommentPattern = regexp.MustCompile(`(?i)--\s*notnull:\s*([^\r\n]+)`)
	stringLiteralPattern  = regexp.MustCompile(`'([^']|'')*'`)
	lineCommentPattern    = regexp.MustCompile(`--[^\r\n]*`)
)

// Validate checks every entry and entity registry declares, plus its
// migration history when registry.Migrations is set, against db's database
// (§7.18). Returns nil when clean, else *core.ValidationErrors carrying every
// violation. Nothing validated leaves a trace: statements are described, never
// executed, inside a shield transaction on the session's connection that is
// always rolled back — TX-001 when the session already has one active.
func Validate(ctx context.Context, db *Db, registry Registry) error {
	return validateAll(ctx, db, registry)
}

// ValidateTypes is the partial form (test harnesses): entries and entities,
// without the migration check (no MIG-030/010/011).
func ValidateTypes(ctx context.Context, db *Db, entries []Entry, entities []reflect.Type) error {
	return validateAll(ctx, db, Registry{Entries: entries, Entities: entities})
}

func validateAll(ctx context.Context, db *Db, registry Registry) error {
	var errs []*core.ValidationError

	if registry.Migrations != nil {
		if err := checkMigrations(ctx, db, registry, &errs); err != nil {
			return err
		}
	}

	tx, err := db.Begin(ctx)
	if err != nil {
		return err
	}
	defer func() { _ = tx.Rollback(ctx) }()

	tableColumns := map[string]map[string]migrations.LiveColumn{}

	for _, entry := range registry.Entries {
		source := entry.Source().Description()
		sqlText, err := entry.Source().SQL()
		if err != nil {
			errs = append(errs, &core.ValidationError{Code: core.CodeOf(err), Source: source, Message: err.Error()})
			continue
		}
		if err := validateStatement(
			ctx, db, tableColumns, source, sqlText, entry.ArgsType(), entry.ResultType(), &errs,
		); err != nil {
			return err
		}
	}

	for _, entityType := range registry.Entities {
		if err := validateEntity(ctx, db, tableColumns, entityType, &errs); err != nil {
			return err
		}
	}

	if len(errs) > 0 {
		return &core.ValidationErrors{Errors: errs}
	}
	return nil
}

// --- registry entries and statement-backed entities -----------------------

// validateStatement is the one statement check (§7.19), shared by registry
// entries and statement-backed entities: PRM-001/002 statically (skipped when
// argsType is nil — a statement entity's own parameters are a loader-time
// concern, ADR-0008 add.2, not this check); the VAL-020/021 lints; then
// describe without executing (VAL-001 on failure); for a query
// (resultType != nil) the result shape through the one mapping pipeline
// (MAP-001/002) and, per described column, the VAL-010/011 checks.
func validateStatement(
	ctx context.Context, db *Db, tableColumns map[string]map[string]migrations.LiveColumn,
	source string, sqlText string, argsType, resultType reflect.Type, errs *[]*core.ValidationError,
) error {
	if argsType != nil {
		checkParameters(sqlText, source, argsType, errs)
	}

	stripped := stripLiteralsAndComments(sqlText)
	if selectStarPattern.MatchString(stripped) {
		*errs = append(*errs, &core.ValidationError{
			Code: "VAL-021", Source: source, Message: "SELECT * is not allowed; list columns explicitly",
		})
	}
	if nonUTCTimestampSQL.MatchString(stripped) {
		*errs = append(*errs, &core.ValidationError{
			Code: "VAL-020", Source: source,
			Message: "current_timestamp/datetime('now') store datetimes without a UTC marker; bind an ISO-8601 Z value instead",
		})
	}

	describer, ok := db.options.Dialect.(core.StatementDescriber)
	if !ok {
		*errs = append(*errs, &core.ValidationError{
			Code: "VAL-001", Source: source, Message: fmt.Sprintf("%T cannot describe statements without executing them", db.options.Dialect),
		})
		return nil
	}

	columns, err := describer.DescribeStatement(ctx, db.connection, sqlText)
	if err != nil {
		*errs = append(*errs, &core.ValidationError{Code: "VAL-001", Source: source, Message: err.Error()})
		return nil
	}

	if resultType == nil {
		return nil
	}

	columnNames := make([]string, len(columns))
	for i, column := range columns {
		columnNames[i] = column.Name
	}
	if _, planErr := db.mapper.Plan(resultType, columnNames, source); planErr != nil {
		appendPlanError(errs, source, planErr)
	}

	notNullOverrides := parseNotNullOverrides(sqlText)
	for _, column := range columns {
		member, err := resolveMember(db, resultType, column.Name)
		if err != nil {
			return err
		}
		if member == nil {
			continue // MAP-001 already reported by Plan above.
		}

		if column.Table != "" && column.OriginColumn != "" {
			info, err := getTableColumnInfo(ctx, db, tableColumns, column.Table, column.OriginColumn)
			if err != nil {
				return err
			}
			if info != nil {
				checkColumnAgainstOrigin(source, column.Name, column.Table, *info, member, db, errs)
				continue
			}
		}

		// Expression column: nullability unknowable — require nullable or the override comment.
		if !member.nullable && !notNullOverrides[strings.ToLower(column.Name)] {
			*errs = append(*errs, &core.ValidationError{
				Code: "VAL-010", Source: source,
				Message: fmt.Sprintf(
					"expression column '%s' has unknown nullability; make the member nullable or add '-- notnull: %s'",
					column.Name, column.Name),
			})
		}
	}
	return nil
}

func checkColumnAgainstOrigin(
	source, columnName, table string, info migrations.LiveColumn, member *resolvedMember, db *Db, errs *[]*core.ValidationError,
) {
	if !db.options.Dialect.IsDeclaredTypeCompatible(info.DeclaredType, member.token) && !db.converter.HasHandler(member.goType) {
		*errs = append(*errs, &core.ValidationError{
			Code: "VAL-011", Source: source,
			Message: fmt.Sprintf(
				"column '%s' is declared %s in %s, incompatible with %s (no handler)",
				columnName, info.DeclaredType, table, member.goType),
		})
	}
	if !info.EffectivelyNotNull() && !member.nullable {
		*errs = append(*errs, &core.ValidationError{
			Code: "VAL-010", Source: source,
			Message: fmt.Sprintf("nullable column '%s' maps to non-nullable %s", columnName, member.goType),
		})
	}
}

// checkParameters is PRM-001/002, statically and both directions (§7.13):
// every placeholder in sqlText must name an exported field of argsType, and
// every exported field must be used by some placeholder.
func checkParameters(sqlText, source string, argsType reflect.Type, errs *[]*core.ValidationError) {
	placeholders := core.FindPlaceholders(sqlText)
	properties := publicFieldNames(argsType)

	for _, placeholder := range placeholders {
		if !containsFold(properties, placeholder) {
			*errs = append(*errs, &core.ValidationError{
				Code: "PRM-001", Source: source,
				Message: fmt.Sprintf("SQL parameter @%s has no property on %s", placeholder, argsType.Name()),
			})
		}
	}
	for _, property := range properties {
		if !containsFold(placeholders, property) {
			*errs = append(*errs, &core.ValidationError{
				Code: "PRM-002", Source: source,
				Message: fmt.Sprintf("property %s.%s is never used by the SQL", argsType.Name(), property),
			})
		}
	}
}

// appendPlanError unwraps a mapping pipeline failure into one ValidationError
// per violation: every error inside a *core.MappingErrors (an entity result
// whose own map failed to load), or the single code/message of a plain
// *core.Error (MAP-001/002 from Plan itself).
func appendPlanError(errs *[]*core.ValidationError, source string, err error) {
	var aggregate *core.MappingErrors
	if errors.As(err, &aggregate) {
		for _, e := range aggregate.Errors {
			*errs = append(*errs, &core.ValidationError{Code: e.Code, Source: source, Message: e.Message})
		}
		return
	}
	var single *core.Error
	if errors.As(err, &single) {
		*errs = append(*errs, &core.ValidationError{Code: single.Code, Source: source, Message: single.Message})
		return
	}
	*errs = append(*errs, &core.ValidationError{Code: "MAP-001", Source: source, Message: err.Error()})
}

// resolvedMember is one result column's target, independent of whether
// resultType is an entity, a DTO, or a scalar (§7.19's three shapes).
type resolvedMember struct {
	goType   reflect.Type
	token    core.ColumnType
	nullable bool
}

// resolveMember finds column's target on resultType: the mapped property by
// column name (entity, case-insensitive), the type itself (scalar, using
// mapping.IsScalarType), or the matching exported field (DTO, using
// mapping.NamesMatch) — the same shape mapping.dtoBindings/entityBindings
// resolve, walked again here because Plan (already run above for MAP-001/002)
// does not expose the per-column origin/nullability detail this check needs.
func resolveMember(db *Db, resultType reflect.Type, column string) (*resolvedMember, error) {
	element := resultType
	if element.Kind() == reflect.Pointer {
		element = element.Elem()
	}

	if metadata.HasMappingDeclarations(element) {
		m, err := db.maps.Load(element)
		if err != nil {
			return nil, nil // Already reported via Plan's own load of the same map.
		}
		for _, property := range m.Properties {
			if strings.EqualFold(property.ColumnName, column) {
				return &resolvedMember{goType: property.Type, token: property.ColumnType, nullable: property.IsNullable}, nil
			}
		}
		return nil, nil // MAP-001 already reported by Plan.
	}

	if mapping.IsScalarType(element, db.converter) {
		return &resolvedMember{
			goType: resultType, token: core.ColumnTypeOf(element), nullable: resultType.Kind() == reflect.Pointer,
		}, nil
	}

	for _, field := range reflect.VisibleFields(element) {
		if !field.IsExported() || field.Anonymous {
			continue
		}
		if mapping.NamesMatch(field.Name, column) {
			return &resolvedMember{
				goType: field.Type, token: core.ColumnTypeOf(field.Type), nullable: field.Type.Kind() == reflect.Pointer,
			}, nil
		}
	}
	return nil, nil // MAP-001 already reported by Plan.
}

// --- entities ---------------------------------------------------------------

// validateEntity is the per-entity check (§7.19): a loader failure
// contributes one ValidationError per violation; procedures/materialized
// views the dialect cannot host are skipped (dormant); a statement entity
// validates like a query with a nil args type; tables and views check
// existence and columns, tables additionally declared types and nullability.
func validateEntity(
	ctx context.Context, db *Db, tableColumns map[string]map[string]migrations.LiveColumn,
	entityType reflect.Type, errs *[]*core.ValidationError,
) error {
	source := entityType.Name()
	m, err := db.maps.Load(entityType)
	if err != nil {
		var aggregate *core.MappingErrors
		if errors.As(err, &aggregate) {
			for _, e := range aggregate.Errors {
				*errs = append(*errs, &core.ValidationError{Code: e.Code, Source: source, Message: e.Message})
			}
			return nil
		}
		return err
	}

	switch m.Kind {
	case core.RelationProcedure:
		if !db.options.Dialect.SupportsProcedures() {
			return nil // dormant on this dialect (capability-gated)
		}
	case core.RelationMaterializedView:
		if !db.options.Dialect.SupportsMaterializedViews() {
			return nil // dormant on this dialect (capability-gated)
		}
	}

	if m.Kind == core.RelationStatement {
		return validateStatement(ctx, db, tableColumns, source+" [Statement]", m.DefiningSQL, nil, entityType, errs)
	}

	columns, err := migrations.ReadLiveColumns(ctx, db.connection, db.options.Dialect, m.RelationName)
	if err != nil {
		return err
	}
	if len(columns) == 0 {
		*errs = append(*errs, &core.ValidationError{
			Code: "VAL-012", Source: source, Message: fmt.Sprintf("relation '%s' does not exist in the database", m.RelationName),
		})
		return nil
	}

	for _, property := range m.Properties {
		info, ok := columns[strings.ToLower(property.ColumnName)]
		if !ok {
			*errs = append(*errs, &core.ValidationError{
				Code: "VAL-013", Source: source,
				Message: fmt.Sprintf("mapped column '%s' does not exist in '%s'", property.ColumnName, m.RelationName),
			})
			continue
		}

		if m.Kind != core.RelationTable {
			continue // views report neither declared types nor nullability reliably (§7.19)
		}

		if !db.options.Dialect.IsDeclaredTypeCompatible(info.DeclaredType, property.ColumnType) && !db.converter.HasHandler(property.Type) {
			*errs = append(*errs, &core.ValidationError{
				Code: "VAL-011", Source: source,
				Message: fmt.Sprintf(
					"column '%s' is declared %s, incompatible with %s (no handler)",
					property.ColumnName, info.DeclaredType, property.Target()),
			})
		}
		if !info.EffectivelyNotNull() && !property.IsNullable {
			*errs = append(*errs, &core.ValidationError{
				Code: "VAL-010", Source: source,
				Message: fmt.Sprintf("nullable column '%s' maps to non-nullable %s", property.ColumnName, property.PropertyName),
			})
		}
	}
	return nil
}

// --- migrations (MIG-030/010/011) -------------------------------------------

func checkMigrations(ctx context.Context, db *Db, registry Registry, errs *[]*core.ValidationError) error {
	var snapshots *migrations.SnapshotSet
	if registry.Snapshots != nil {
		set, err := migrations.SnapshotsFromFS(registry.Snapshots)
		if err != nil {
			code := core.CodeOf(err)
			if code == "" {
				return err
			}
			*errs = append(*errs, &core.ValidationError{Code: code, Source: "migrations", Message: err.Error()})
			return nil
		}
		snapshots = set
	}

	runner := migrations.NewRunner(db.pool, db.options.Dialect, db.maps, registry.Migrations, snapshots)
	entries, err := runner.Status(ctx)
	if err != nil {
		code := core.CodeOf(err)
		if code == "" {
			return err
		}
		*errs = append(*errs, &core.ValidationError{Code: code, Source: "migrations", Message: err.Error()})
		return nil
	}

	for _, entry := range entries {
		var code, message string
		switch entry.State {
		case migrations.Pending:
			code, message = "MIG-030", "pending — apply migrations before starting the application"
		case migrations.Drifted:
			code, message = "MIG-010", "applied with a different checksum than the code renders"
		case migrations.Unknown:
			code, message = "MIG-011", "applied in the database but unknown to the code"
		default:
			continue
		}
		*errs = append(*errs, &core.ValidationError{
			Code: code, Source: fmt.Sprintf("V%04d %s", entry.Version, entry.ObjectName), Message: message,
		})
	}
	return nil
}

// --- introspection helpers ---------------------------------------------------

// getTableColumnInfo is table's column, through migrations.ReadLiveColumns
// (the one introspection helper, CODING-STANDARD §8), cached per relation
// across the whole Validate call.
func getTableColumnInfo(
	ctx context.Context, db *Db, cache map[string]map[string]migrations.LiveColumn, table, column string,
) (*migrations.LiveColumn, error) {
	columns, ok := cache[table]
	if !ok {
		var err error
		columns, err = migrations.ReadLiveColumns(ctx, db.connection, db.options.Dialect, table)
		if err != nil {
			return nil, err
		}
		cache[table] = columns
	}
	info, ok := columns[strings.ToLower(column)]
	if !ok {
		return nil, nil
	}
	return &info, nil
}

func publicFieldNames(t reflect.Type) []string {
	if t.Kind() == reflect.Pointer {
		t = t.Elem()
	}
	var names []string
	for _, field := range reflect.VisibleFields(t) {
		if field.IsExported() && !field.Anonymous {
			names = append(names, field.Name)
		}
	}
	return names
}

func containsFold(haystack []string, needle string) bool {
	for _, candidate := range haystack {
		if strings.EqualFold(candidate, needle) {
			return true
		}
	}
	return false
}

func stripLiteralsAndComments(sqlText string) string {
	withoutLiterals := stringLiteralPattern.ReplaceAllString(sqlText, "''")
	return lineCommentPattern.ReplaceAllString(withoutLiterals, "")
}

func parseNotNullOverrides(sqlText string) map[string]bool {
	overrides := map[string]bool{}
	for _, match := range notNullCommentPattern.FindAllStringSubmatch(sqlText, -1) {
		for _, name := range strings.Split(match[1], ",") {
			overrides[strings.ToLower(strings.TrimSpace(name))] = true
		}
	}
	return overrides
}
